Structure Generation Fixes

Dariensg opened the pull request:
-Made Strider Altars check for air above the lava so that they generate in the open and not in the wall
-Created new Generation Validity checks for all other structures that look for a fairly flat and open area of land to spawn in with enough vertical space for the structure
-Changed Structure spacing and separation to match the new generation checks to get reasonable numbers of structures
-Removed x and z coordinate randomization of structures as it was unnecessary
-Had several headaches but finally got it working

Lots of credit goes to TelepathicGrunt for all the help with this one

Dariensg committed:
Initial Code

-Added in code to check for space for strider altars
-Added preliminary code to other structure types to check if area is big enough

Dariensg committed:
Bastion New Gen Check Test

-Added new horizontal generation check to Bastion Outpost for testing
-Using locate on bastion outpost while using this generation stops all ticks and doesn't seem to ever resume

Dariensg committed:
Fix Infinite Loop

-Fixed Infinite loop in Bastion Outpost

Dariensg committed:
Cleaned Up Flat Land Check

-Cleaned up the check for a block with air above it
-Made sure it checks for solid blocks, not air

Dariensg committed:
Made SharedSeedRandom Inline

-Made both SharedSeedRandom instances inline instead of being fields

Dariensg committed:
Tweaked Values

-Tweaked values to decrease rarity of outposts
-Removed randomization of x and z coordinates of structure

Dariensg committed:
Added New Gen to All Structures

-Added new Generation Checks to SSV and GSC Ruins
-Tweaked some values to get the right frequency of structure spawns

Dariensg committed:
Merge branch 'master-indev' of https://github.com/infernalexp/Infernal-Expansion into structure-fixes

Dariensg committed:
Removed Strider Altar X,Z Randomization

-Removed x and z coordinate randomization of strider altars

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java:
```diff
@@ -35,32 +43,114 @@ public BastionOutpostStructure(Codec<NoFeatureConfig> codec) {
+                for (int curZ = posZ - 8; curZ <= posZ + 8; curZ += 8) {
+
+                    // Starts 5 blocks below to check for solid land in each column
+                    BlockPos.Mutable mutable = new BlockPos.Mutable();
+                    mutable.setPos(curX, posY - 5, curZ);
```
```suggestion
                    BlockPos.Mutable mutable = new BlockPos.Mutable(curX, posY - 5, curZ);
```

Hellion-Games replied:
The thing about moving func_240363_a_ to a super class is that each structure type uses different values for the checks, so I'd still just have to override it in each subclass specific structure anyway. I could move the getYPos method, but that's about the only one that makes sense.

SwanX1 replied:
Couldn't you just add new arguments for those values?

Hellion-Games replied:
Fair enough, but there's a lot of them. There's how far away to go from the initial point, how much to increment after each sample, how vertically to check for land, and how vertically to check for air above the structure. That feels like a bit much but I can do it if you think I should.

SwanX1 replied:
Hmm... if it's that much, then don't implement it in the superclass. (also this thread is in the wrong place, this is about a suggestion for the code on line 70 LMAO)

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java:
```diff
@@ -35,32 +43,114 @@ public BastionOutpostStructure(Codec<NoFeatureConfig> codec) {
+        }
+
+        return true;
+    }
+    public int getYPos(ChunkGenerator chunkGenerator, int x, int z, SharedSeedRandom random) {
```
```suggestion
    }
    
    public int getYPos(ChunkGenerator chunkGenerator, int x, int z, SharedSeedRandom random) {
```

SwanX1 replied:
Documenting this method would be very helpful for future contributors.

SwanX1 replied:
Shouldn't this method be static?

SwanX1 commented on src/main/java/com/nekomaster1000/infernalexp/world/gen/structures/BastionOutpostStructure.java:
```diff
@@ -35,32 +43,114 @@ public BastionOutpostStructure(Codec<NoFeatureConfig> codec) {
+                    }
+                }
+            }
+        }
+        else {
```
```suggestion
        } else {
```

SwanX1 reviewed (changes requested):
the same comments from `BastionOutpostStructure::func_230363_a_` and `BastionOutpostStructure::getYPos` apply to `GlowstoneCanyonRuinStructure`, `SoulSandValleyRuinStructure`, and `StriderAltarStructure`.

I'd also recommend refactoring the code to make all the previously mentioned classes inherit a class that has their shared code, like `func_230363_a_`, `getYPos`, and the inner class `Start`.
The inheritance would be as such
 - `SpecificStructure`
   - `SolidGroundSpawningStructure<Config>` <- new class
     - `IEStructure<Config>`

Dariensg committed:
Fix PR Suggestions

-Moved getYPos completely to superclass IEStructure
-Documented getYPos
-Fixed minor formatting issues

Dariensg commented:
> 
> 
> Please fix the compilation issues
Weird. It launches fine for me from my dev environment and the error that the build failed on was from a mixin with ItemModelsProperties it looked like. I'm trying re-running it.

Dariensg committed:
Merge branch 'master-indev' of https://github.com/infernalexp/Infernal-Expansion into structure-fixes

Dariensg committed:
Merge branch 'master-indev' of https://github.com/infernalexp/Infernal-Expansion into structure-fixes

SwanX1 commented:
Merged in commit d18021f9b0b5f7733d4536c98bed09dfb458473f, pull request doesn't show as merged due to me removing the last two merge commits to avoid unnecessary commit spam.
